### PR TITLE
Mock connection for ProtocolAdapter

### LINKED ISSUE #742: How to test adapter.connection calls
Some instruments call the adapter's `connection`property, like #697 which makes ProtocolAdapter fail. Instruments should not be adjusted to make the ProtocolAdapter work.

The problem is, that you are not supposed to use normally the adapter's connection directly, because the `connection` depends on the specific adapter type. If an instrument calls the connection, the ProtocolAdapter fails. Instruments should not be adjusted to make the ProtocolAdapter work.

Two solutions:

- A generic mock connection seems good, because that would care for many use cases. Otherwise we had to add several methods (clear, close, timeout property, get_stb...) individually.
- Or we could add timeout as a property to the `Adapter`, which will call its connection depending on the Adapter type (ProtocolAdapter stores it as a variable). Similarly the "read_stb" and `clear` etc.

### PR DESCRIPTION
**Pull request by bmoneke**
Closes #742 

The connection of the ProtocolAdapter is a MagicMock.
You can set attribute return values and method return values via a dictionary (which you can set from `expected_protocol` as well).
Such that you can preset your connection in order to let tests pass (see Lecroy test).

### CODE REVIEW

**Review by msmttchr — APPROVED** [review]
Looks good to me. Just a minor suggestion on a comment.

File: pymeasure/test.py [hunk 1]
```diff
@@ -47,10 +49,13 @@ def expected_protocol(instrument_cls, comm_pairs, **kwargs):
         'None' indicates that a pair member (command or response) does not
         exist, e.g. `(None, 'RESP1')`. Commands and responses are without
         termination characters.
+    :param dict connection_attributes: Dictionary of attributes of the connection.
+    :param dict connection_methods: Pairs of method names and return values.
```
[the remark this answers is not in the document]
Comment (bmoneke) [reply pymeasure/test.py:53]: I wrote it between param and the name, but yes, explicit is better.

**Comment by bilderbuchi** [comment]
I'm on the phone, so can't review right now, but one design question: does the magicmock signal/do we notice if a test made an unexpected call we did not expect? Similar to the way we notice if a protocoladapter has comm pairs left over when it goes out of scope?
Or is that not a situation worth guarding against?

**Comment by bmoneke** [comment]
> One design question: does the magicmock signal/do we notice if a test made an unexpected call we did not expect? Similar to the way we notice if a protocoladapter has comm pairs left over when it goes out of scope? Or is that not a situation worth guarding against?

There is no check, whether the mockconnection received any unexpected calls.
There are not many calls to the connection anyway. In these cases you have to make the right tests (verify that the connection got called as expected etc.).
Most of the time, implementors should use read/write. It is mainly for edge cases (with that timeout or the chunk_size of the Lecroy...), such that the instruments do not have to be adjusted to the tests, but the tests to the instruments.

From the implementation, you can limit the possible calls and can test that they are done in the appropriate number. That is lot less work than to implement some univeral testing scheme.

**Review by bilderbuchi — APPROVED** [review]
LGTM, nice useful feature!
